wagtail/permission_policies/pages.py

```diff
@@ -15,6 +17,15 @@
 
 
 class PagePermissionPolicy(BasePermissionPolicy):
+    perm_cache_name = "_page_perm_cache"
+
+    def get_all_permissions_for_user(self, user):
+        if not user.is_active or user.is_anonymous or user.is_superuser:
+            return set()
+        return set(
+            GroupPagePermission.objects.filter(group__user=user).select_related("page")
+        )
```

Comment:
If you don't like the fact that we're querying and caching **all** `GroupPagePermission` instances for the user during the request-response cycle, I have news for you:

We've been doing the same thing in `UserPagePermissionsProxy`:

https://github.com/wagtail/wagtail/blob/92a3cc246ed60ed8c04e047da9c33fae7fb7f15c/wagtail/models/__init__.py#L2898-L2908

and it's even worse, because there are places in the admin where we instantiate a new `UserPagePermissionsProxy` object instead of reusing the cached instance in the template context (because we don't always have access to the context). This means the query is duplicated quite a few times in a single request-response cycle.

Reply:
Yep, don't worry - I was aware of that :-) When I came up with the `annotate_with_permissions` strategy, I had it in mind as an improvement over the status quo - my thinking was: if UserPagePermissionProxy objects are typically only kept around for the duration of iterating over one queryset, we should at least optimise the permission lookups to be localised to that queryset. I'm pretty sure that the gains we make from caching that data more effectively, on a per-request level, will trump those improvements, though.

Reply:
Ahh, yep. Considering how we interact with page permissions in different places when rendering the admin views (e.g. the sidebar menu needs to calculate the explorable pages), I think splitting the permissions query won't be a big benefit. Yes, we will fetch fewer permission records, but we'll need to make separate queries for other use cases. That potentially would make the code harder to follow as well, since we'd have to write the filter that narrows down the permission query for each case (or at least based on a particular page instance).

It's pretty unlikely for a user object to have a large number of permission records to the point where this becomes an issue, considering that we've been doing the same (and worse) via `UserPagePermissionsProxy` and haven't heard any complaints (not that I know of).